src/tiledb/cloud/geospatial/ingestion.py

```diff
@@ -0,0 +1,331 @@
+                pipeline |= pdal.Reader.las(filename=i)
+                pipeline |= pdal.Writer.tiledb(array_name=output,
+                                            **kwargs
+                                            )
+
+        pipeline.execute()
+
+    def ingest_raster_udf(
+        inputs: Tuple[tuple[(int, int), (int, int)], tuple[str, ...]],
+        output: str,
+        *args: Any,
+        **kwargs,
+    ):
+        """Internal udf that ingests server side batch of raster files
+        into tiledb arrays using Rasterio API
+
+        :param io_uris: raster input file names
+        :param output: output TileDB array name
+        """
+
+        import rasterio
+        import rasterio.merge
+
+        append = kwargs.get("append", False)
+
+        if not append:
+            from rasterio.transform import from_origin
+
+            from tiledb import filter
+
+            compressor = kwargs.get("compressor", None)
+            extents = kwargs.pop("extents", None)
+            tile_size = kwargs.get("tile_size", 1024)
+            res = kwargs.pop("res", None)
+
+            if compressor and extents and res:
+                compressor_args = dict(compressor)
+                compressor_name = compressor_args.pop("_name")
+                if compressor_name:
+                    compressor_args = {
+                        k: None if not v else v for k, v in compressor_args.items()
+                    }
+                    compression_filter = vars(filter).get(compressor_name)(
+                        **compressor_args
+                    )
+                    compression_level = compression_filter.level
+
+                else:
+                    raise ValueError
+
+                # create empty array, inputs is a single reference file when creating the array
+                with rasterio.open(inputs) as src:
+                    dst_profile = src.profile.copy()
+                    dst_profile.update({
+                        "driver": "TileDB",
+                        "transform": from_origin(extents.minx, extents.maxy, *res),
+                        "width": math.floor((extents.maxx - extents.minx) / res[0]),
+                        "height": math.floor((extents.maxy - extents.miny) / res[1]),
+                        "blockysize": tile_size,
+                        "blockxsize": tile_size,
+                    })
+                    with rasterio.open(output, "w",
+                                       COMPRESSION=compressor_name[:-6].upper(), # remove trailing `Filter`
+                                       COMPRESSION_LEVEL=compression_level,
+                                       **dst_profile,):
+                        pass
+        else:
+            # srcs and dst have same number of bands
+            resampling = rasterio.enums.Resampling[kwargs.pop("resampling", "bilinear").lower()]
+            with rasterio.open(output, mode="r+") as dst:
+                for b in range(dst.count):
```

Comment:
@normanb I don't understand why the inputs aren't the outer loop. I'm probably missing something conceptually.

Reply:
I did this because you can technically have different block sizes per band in GDAL, however in TileDB you cannot so I will move the band loop up a level and reduce the number of times the datasets are opened.